Cleanup ChannelId handling of basic methods

jasontedor opened the pull request:
Motiviation:

Simplify implementation of compareTo/equals/hashCode for ChannelIds.

Modifications:

We simplfy the hashCode implementation for DefaultChannelId by not making it random, but making it based on the underlying data. We fix the compareTo implementation for DefaultChannelId by using lexicographic comparison of the underlying data array. We fix the compareTo implementation for CustomChannelId to avoid the possibility of overflow.

Result:

Cleaner code that is easier to maintain.

Closes #6371

Scottmitch commented on transport/src/main/java/io/netty/channel/ChannelId.java:
```diff
@@ -43,7 +43,7 @@
  * </p>
  */
-public interface ChannelId extends Serializable, Comparable<ChannelId> {
+public interface ChannelId<T extends ChannelId> extends Serializable, Comparable<T> {
```
this is an API change. I don't think we can do this in a minor release. Can you revert this change please?

normanmaurer replied:
+1 .. Please revert

Scottmitch commented on transport/src/main/java/io/netty/channel/DefaultChannelId.java:
```diff
@@ -247,25 +247,36 @@ public int hashCode() {
+        int len2 = o.data.length;
+        int len = Math.min(len1, len2);
+
+        int k = 0;
```
consider just using a for loop `for (int k = 0; k < len; ++k)`

Scottmitch commented on transport/src/main/java/io/netty/channel/DefaultChannelId.java:
```diff
@@ -247,25 +247,36 @@ public int hashCode() {
     public String toString() {
         return asShortText();
     }
+
```
nit: kill line?

Scottmitch commented on transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java:
```diff
@@ -62,4 +58,5 @@ public boolean equals(Object obj) {
     public String toString() {
         return "embedded";
     }
+
```
nit: kill line?

Scottmitch commented on transport/src/main/java/io/netty/channel/DefaultChannelId.java:
```diff
@@ -247,25 +247,36 @@ public int hashCode() {
+            byte x = data[k];
+            byte y = o.data[k];
+            if (x != y) {
+                return x - y;
```
consider adding a comment that overflow won't occur because the arithmetic is promoted to `int`.

Scottmitch commented on transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java:
```diff
@@ -17,21 +17,20 @@
-        return -1;
+    public int compareTo(CustomChannelId o) {
+        // do not subtract for comparison, it could overflow
+        return id < o.id ? -1 : (id == o.id ? 0 : 1);
```
consider adding a `compare` to MathUtil for `int`. That way when we upgrade our java support it will be easier to identify and replace with the JDK methods.

fenik17 commented:
IMHO, for  lexicographic comparison more suitable compare bytes as unsigned.

jasontedor commented:
Thanks for the feedback @Scottmitch and @normanmaurer. I have pushed a commit addressing your comments. I also agree with the suggestion of @fenik17 to do the lexicographic comparison treating the bytes as unsigned, so I included that in that commit as well.

normanmaurer commented on transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java:
```diff
@@ -40,12 +40,12 @@ public String asLongText() {
         if (o instanceof EmbeddedChannelId) {
             return 0;
+        } else {
+            return asLongText().compareTo(o.asLongText());
```
please revert this as the `else`is not needed or make this:

```java
return o instanceof EmbeddedChannelId ? 0 : asLongText().compareTo(o.asLongText());
```

normanmaurer commented on transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java:
```diff
@@ -16,22 +16,26 @@
-            return id - ((CustomChannelId) o).id;
+            final int otherId = ((CustomChannelId)o).id;
+            return MathUtil.compare(id, otherId);
+        } else {
```
`} else {` is not needed. Please remove

jasontedor commented:
Thanks again @normanmaurer, I pushed another commit.

normanmaurer commented on transport/src/main/java/io/netty/channel/DefaultChannelId.java:
```diff
@@ -247,21 +247,35 @@ public int hashCode() {
+                }
+            }
+            return len1 - len2;
+        } else {
```
remove else block

jasontedor replied:
I pushed a commit that does this.

normanmaurer commented on transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java:
```diff
@@ -40,7 +40,7 @@ public String asLongText() {
 
     @Override
-    public int compareTo(ChannelId o) {
+    public int compareTo(final ChannelId o) {
```
revert ?

jasontedor replied:
Why? The `final` is fine here. This is an improvement that breaks nothing.

normanmaurer replied:
Just because it not really adds anything to the patch IMHO... that said I don't care too much

jasontedor replied:
The goal of the patch was to clean up equals/compareTo/hashCode methods on the `ChannelId` implementations. :smile:

trustin replied:
I think it's always good to clean the code up while making modifications as long as it doesn't drain too much cognitive load. :-)

normanmaurer commented on transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java:
```diff
@@ -16,22 +16,26 @@
         if (o instanceof CustomChannelId) {
-            return id - ((CustomChannelId) o).id;
+            final int otherId = ((CustomChannelId)o).id;
+            return MathUtil.compare(id, otherId);
```
just merge these two lines ?

jasontedor replied:
I pushed a commit that does this.

jasontedor commented:
Thanks @normanmaurer. I've responded to your latest round of feedback.

johnou commented on common/src/main/java/io/netty/util/internal/MathUtil.java:
```diff
@@ -65,6 +65,20 @@ public static boolean isOutOfBounds(int index, int length, int capacity) {
+     *         {@code -1} if {@code x < y}; and
+     *         {@code 1} if {@code x > y}
+     */
+    public static int compare(int x, int y) {
```
do we need to introduce this? can't we use java.lang.Integer#compare?

normanmaurer replied:
We can't as we need to still support java6

johnou replied:
ow, for some reason I thought Netty 4.0.x was java6 and Netty 4.1.x was java7.

jasontedor replied:
No, it's 1.6 in the [pom.xml](https://github.com/netty/netty/blob/c92565d5c7000d2bd74355fe18cbdf98bfc5ba0d/pom.xml#L625).

normanmaurer commented:
@Scottmitch PTAL and pull in when satisfied. 

@jasontedor thanks again. Great work!

jasontedor commented:
Thanks @normanmaurer. I squashed.

Scottmitch commented:
4.1 (d59b484)

Thanks @jasontedor !

Scottmitch commented:
Also 4.1 (544d771)
